openstudiocore/src/model/MeterCustom.cpp

```diff
@@ -0,0 +1,298 @@
+  }
+
+  MeterCustom_Impl::MeterCustom_Impl(const openstudio::detail::WorkspaceObject_Impl& other,
+                                     Model_Impl* model,
+                                     bool keepHandle)
+    : ModelObject_Impl(other, model, keepHandle)
+  {
+    OS_ASSERT(other.iddObject().type() == MeterCustom::iddObjectType());
+  }
+
+  MeterCustom_Impl::MeterCustom_Impl(const MeterCustom_Impl& other,
+                                     Model_Impl* model,
+                                     bool keepHandle)
+    : ModelObject_Impl(other,model,keepHandle)
+  {}
+
+  // This logically doesn't produce any output variables
+  const std::vector<std::string>& MeterCustom_Impl::outputVariableNames() const
+  {
+    static std::vector<std::string> result;
+    if (result.empty()){
+    }
+    LOG(Info, "Meter:Custom does not produce any output variables.");
+    return result;
+  }
+
+  IddObjectType MeterCustom_Impl::iddObjectType() const {
+    return MeterCustom::iddObjectType();
+  }
+
+  // fuel Type
+  boost::optional<std::string> MeterCustom_Impl::fuelType() const {
+    return getString(OS_Meter_CustomFields::FuelType,true);
+  }
+
+  bool MeterCustom_Impl::setFuelType(boost::optional<std::string> fuelType) {
+    bool result(false);
+    if (fuelType) {
+      result = setString(OS_Meter_CustomFields::FuelType, fuelType.get());
+    }
+    else {
+      resetFuelType();
+      result = true;
+    }
+    return result;
+  }
+
+  void MeterCustom_Impl::resetFuelType() {
+    bool result = setString(OS_Meter_CustomFields::FuelType, "");
+    OS_ASSERT(result);
+  }
+
+
+  // Add a new (Key, Var) group
+  bool MeterCustom_Impl::addKeyVarGroup(const std::string& keyName, const std::string& outputVariableorMeterName) {
+    WorkspaceExtensibleGroup eg = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+
+    bool temp = eg.setString(OS_Meter_CustomExtensibleFields::KeyName, keyName);
+    bool ok = eg.setString(OS_Meter_CustomExtensibleFields::OutputVariableorMeterName, outputVariableorMeterName);
+
+    if (temp) {
+      temp = ok;
+    }
+
+    if (!temp) {
+      getObject<ModelObject>().eraseExtensibleGroup(eg.groupIndex());
+      return temp;
+    }
+    return temp;
+  }
+
+  // Remove the (Key, Var) group at given index
+  bool MeterCustom_Impl::removeKeyVarGroup(unsigned groupIndex) {
+    bool ok = false;
+    unsigned numberofDataPairs = numExtensibleGroups();
+    // Only delete if the index is smaller than the number of groups
+    if (groupIndex < numberofDataPairs) {
+      getObject<ModelObject>().eraseExtensibleGroup(groupIndex);
+      ok = true;
+    }
+
+    return ok;
+  }
+
+  // Remove all the (Key, Var) groups
+  void MeterCustom_Impl::removeAllKeyVarGroups() {
+    getObject<ModelObject>().clearExtensibleGroups();
+  }
+
+  // Return a vector of (Key, Var) pairs
+  std::vector< std::pair<std::string, std::string> > MeterCustom_Impl::keyVarGroups() {
```

Comment:
Option 1:
```
meter_custom.addKeyVarGroup(key, var)
meter_custom.keyVarGroups.each do |key_var|
    # key_var is a std::pair<std::string, std::string>  
    key = key_var.first
    var = key_var.second
end
```

Option 2:
```
group = OpenStudio::Model::MeterCustomKeyVarGroup.new(key, var)
meter_custom.addKeyVarGroup(group)
meter_custom.keyVarGroups.each do |group|
    # group is a MeterCustomKeyVarGroup  
    key = group.key
    var = group.var
end
```

Option 3:
```
meter = OpenStudio::Model::Meter(model)
meter_custom.addVariable(meter)
variable = OpenStudio::Model::OutputVariable(var, model)
meter_custom.addVariable(variable)
meter_custom.meterOrVariables.each do |object|
    # object is a ModelObject
    if object.to_Meter.is_initialized?
        meter = object.to_Meter.get
    else
        variable = object.to_Variable.get
    end
end
```

I think Option 2 is a bit more clear than Option 1.  Option 3 does seem unnecessarily complicated, especially since it would require a lot of extra output.  It does seem like we should rename Meter to OutputMeter to avoid confusion.  We can create a SWIG alias for Meter -> OutputMeter.  

Finally, when you make a MeterCustom do you have to make a corresponding OutputMeter to write it to the file?  Our current Meter object (really OutputMeter) doesn't support arbitrary meter names so that might not work?